README.md
# JWT Access Token Authorization Middleware _(@digitalbazaar/authorize-access-token-middleware)_

[![Node.js CI](https://github.com/digitalbazaar/authorize-access-token-middleware/workflows/Node.js%20CI/badge.svg)](https://github.com/digitalbazaar/authorize-access-token-middleware/actions?query=workflow%3A%22Node.js+CI%22)
[![NPM Version](https://img.shields.io/npm/v/@digitalbazaar/authorize-access-token-middleware.svg)](https://npm.im/@digitalbazaar/authorize-access-token-middleware)

> An opinionated Express-like middleware to protect API endpoints via OAuth 2.0 access tokens.

## Table of Contents

- [Background](#background)
- [Security](#security)
- [Install](#install)
- [Usage](#usage)
- [Contribute](#contribute)
- [Commercial Support](#commercial-support)
- [License](#license)

## Background

This is primarily intended for app authorization (for applications using the
`client_credentials` OAuth 2 grant type), rather than for user (`sub`ject)
authentication (for that, Passport is more useful).

Assumptions:

* Access tokens will be passed via HTTP headers, using `Bearer <token>` scheme.

Not supported in v1.0:

* Optional authorization (that is, if you use this middleware, an access token
  is required).
* Automatically decrypting access tokens (although you can decrypt them in the `verify` callback).
* DPoP / proof of possession header functionality.

## Security

TBD

## Install

- Node.js >= 16 is required.

To install locally (for development):

```
git clone https://github.com/digitalbazaar/authorize-access-token-middleware.git
cd authorize-access-token-middleware
npm install
```

## Usage

### Importing

```js
import {authorizeAccessToken} from '@digitalbazaar/authorize-access-token-middleware';
```

### Adding authorization to an api route

```js
app.post('/example/api/endpoint',
  authorizeAccessToken({
    // OAuth2 scope required for this endpoint
    requiredScope: 'my.custom.scope',

    // Optional list of allowed issuers of tokens.
    // If missing, issuer validation must be performed manually
    // in validateClaims() callback.
    validIssuers: ['https://issuer.example.com'],

    verify: async ({token}) => {
      // REQUIRED verify callback must:
      // 1. Decode the claims (decrypting the token, if encrypted)
      // 2. Verify key id, algorithm and signature (using a remote KMS or similar)
      // 3. Return the token payload (claims)
    },

    validateClaims: async ({claims}) => {
      // Optional custom claim validation callback (for example, you can
      // validate the `aud`ience claim). Expected to throw errors as appropriate.
    },

    decorateError: async ({errorResponse}) => {
      // Optional callback to decorate/add to the error response.
      // By default, error responses follow the OAuth 2.0 error response format
      // @see https://datatracker.ietf.org/doc/html/rfc6749#section-5.2
      // This is typically used to add other application-specific fields to the JSON error response
      errorResponse.appSpecificErrorCode = '1234';
      return errorResponse;
    },

    // Optional advanced override (an lru-memoize instance)
    claimsCache: new LruCache({max: 100, maxAge: 5000}),

    // Optional logger object (such as console, pino, winston, and so on)
    logger: console
  }),
  (req, res, next) => {
    // ... continue with your route handling as usual
    // `req.clientMetadata` will be set with
    // the decoded and validated claims from the token
  }
)
```

## Contribute

See [the contribute file](https://github.com/digitalbazaar/bedrock/blob/master/CONTRIBUTING.md)!

PRs accepted.

If editing the Readme, please conform to the
[standard-readme](https://github.com/RichardLitt/standard-readme) specification.

## Commercial Support

Commercial support for this library is available upon request from
Digital Bazaar: support@digitalbazaar.com

## License

[New BSD License (3-clause)](LICENSE) © Digital Bazaar


CHANGELOG.md
# authorize-access-token-middleware Changelog

## 2.0.1 - 2023-02-02

### Fixed
- Fix `node.js >= 16` requirement in `README.md`.

## 2.0.0 - 2023-02-01

### Changed
- **BREAKING**: Convert to module (ESM).
- **BREAKING**: Drop support for node <= 14.
- Update `@digitalbazaar/lru-memoize` to v3.0.0.
- Update dev deps to latest.

### Removed
- Remove unused dep `lru-cache@6.0`.

## 1.0.0 - 2021-08-23

### Added
- Initial commit.


tests/test.js
/*!
 * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
 */
const bedrock = require('bedrock');
require('bedrock-mongodb');
require('bedrock-tokenizer');
require('bedrock-https-agent');

require('bedrock-test');
bedrock.start();


test/main.spec.js
/*!
 * Copyright (c) 2021-2023 Digital Bazaar, Inc. All rights reserved.
 */
import * as JWT from '@digitalbazaar/minimal-jwt';
import {_assertScope} from '../lib/authorizeAccessToken.js';
import {authorizeAccessToken} from '../lib/index.js';
import chai from 'chai';
import chaiHttp from 'chai-http';
import crypto from 'crypto';
import express from 'express';

chai.use(chaiHttp);
chai.should();
const {expect} = chai;

const SECRET = '<the-best-kept-secret>';
const MOCK_CLIENT_ID = 's6BhdRkqt3';

const INVALID_MOCK_ACCESS_TOKEN = 'eyJhbGciOiJFUzI1NiIsImtpZCI6Ijc3In0' +
  '.eyJpc3MiOiJodHRwOi8vYXV0aG9yaXphdGlvbi1zZXJ2ZXIuZXhhbXBsZS5jb20iLCJzdW' +
  'IiOiJfX2JfYyIsImV4cCI6MTU4ODQyMDgwMCwic2NvcGUiOiJjYWxlbmRhciIsImF1ZCI6I' +
  'mh0dHBzOi8vY2FsLmV4YW1wbGUuY29tLyJ9.nNWJ2dXSxaDRdMUKlzs-cYI' +
  'j8MDoM6Gy7pf_sKrLGsAFf1C2bDhB60DQfW1DZL5npdko1_Mmk5sUfzkiQNVpYw';

async function signFn({data}) {
  return crypto.createHmac('sha256', Buffer.from(SECRET)).update(data).digest();
}

async function verifyFn({alg, kid, data, signature}) {
  if(alg === 'HS256') {
    const expectedSignature = await signFn({data});

    return crypto.timingSafeEqual(
      Buffer.from(signature),
      Buffer.from(expectedSignature)
    );
  } else {
    throw new Error(`Key "${kid}" is not supported.`);
  }
}

describe('authorizeAccessToken()', () => {
  const app = express();
  app.post('/api/example',
    authorizeAccessToken({
      // scope required for this endpoint
      requiredScope: 'read write',
      validIssuers: ['https://issuer.example.com'],
      verify: async ({token}) => {
        return (await JWT.verify({jwt: token, verifyFn})).payload;
      },
      // eslint-disable-next-line no-unused-vars
      validateClaims: async ({claims}) => {
        // perform custom validation on access token claims here
      },
      // logger: console
    }),
    // eslint-disable-next-line no-unused-vars
    (req, res, next) => {
      res.json(
        req.clientMetadata
      );
    }
  );

  let requester;
  let validAccessToken;
  let expiredAccessToken;
  let invalidIssuerToken;

  before(async () => {
    const now = Date.now();
    validAccessToken = await JWT.sign({
      header: {typ: 'JWT', alg: 'HS256', kid: '194B72684'},
      payload: {
        iss: 'https://issuer.example.com',
        iat: now / 1000,
        exp: now / 1000 + 1000,
        client_id: MOCK_CLIENT_ID,
        scope: 'read write'
      },
      signFn
    });

    expiredAccessToken = await JWT.sign({
      header: {typ: 'JWT', alg: 'HS256', kid: '194B72684'},
      payload: {
        iss: 'https://issuer.example.com',
        iat: Date.now() / 1000,
        exp: (Date.now() / 1000) - 10,
        client_id: MOCK_CLIENT_ID,
        scope: 'read write'
      },
      signFn
    });

    invalidIssuerToken = await JWT.sign({
      header: {typ: 'JWT', alg: 'HS256', kid: '194B72684'},
      payload: {
        iss: 'https://invalid-issuer.com',
        iat: Date.now() / 1000,
        exp: (Date.now() / 1000) + 10,
        client_id: MOCK_CLIENT_ID,
        scope: 'read write'
      },
      signFn
    });

    requester = chai.request(app).keepOpen();
  });

  after(async () => {
    requester.close();
  });

  it('should error if missing authorization header', async () => {
    const res = await requester.post('/api/example')
      .send({});
    expect(res).to.have.status(400);
    expect(res).to.be.json;
    expect(res.body.error).to.equal('invalid_request');
    expect(res.body.error_description)
      .to.equal('Missing "authorization" header.');
  });

  it('should error if authorization header is invalid', async () => {
    const res = await requester.post('/api/example')
      .set('authorization', `notBearer ${validAccessToken}`)
      .send({});
    expect(res).to.have.status(400);
    expect(res).to.be.json;
    expect(res.body.error).to.equal('invalid_request');
    expect(res.body.error_description)
      .to.equal('Invalid "authorization" header.');
  });

  it('should error if cannot verify access token', async () => {
    const res = await requester.post('/api/example')
      .set('authorization', `Bearer ${INVALID_MOCK_ACCESS_TOKEN}`)
      .send({});
    expect(res).to.have.status(401);
    expect(res).to.be.json;

    expect(res.body.error).to.equal('invalid_token');
    expect(res.body.error_description).to.equal(
      'Error verifying the access token.');
  });

  it('should error if issuer is not authorized', async () => {
    const res = await requester.post('/api/example')
      .set('authorization', `Bearer ${invalidIssuerToken}`)
      .send({});
    expect(res).to.have.status(401);
    expect(res).to.be.json;

    expect(res.body.error).to.equal('invalid_token');
    expect(res.body.error_description).to.equal(
      'Token issuer "https://invalid-issuer.com" is not authorized.');
  });

  it('should error if token expired', async () => {
    const res = await requester.post('/api/example')
      .set('authorization', `Bearer ${expiredAccessToken}`)
      .send({});
    expect(res).to.have.status(401);
    expect(res).to.be.json;

    expect(res.body.error).to.equal('invalid_token');
    expect(res.body.error_description).to.equal('Access token has expired.');
  });

  it('should successfully authorize access token', async () => {
    const res = await requester.post('/api/example')
      .set('content-type', 'application/json')
      .set('authorization', `Bearer ${validAccessToken}`)
      .send({});
    expect(res).to.have.status(200);
    expect(res.body.iss).to.equal('https://issuer.example.com');
  });
});

describe('_assertScope', () => {
  it('should throw on missing params', () => {
    expect(() => _assertScope())
      .to.throw();
  });

  it('should throw on token scope/required scope mismatch', () => {
    let error;
    try {
      _assertScope({
        tokenScope: 'read', requiredScope: 'write'
      });
    } catch(e) {
      error = e;
    }

    expect(error).to.exist;
    expect(error.error).to.equal('insufficient_scope');
    expect(error.error_description).to
      .equal('Access denied for token scope "read".');
    expect(error.statusCode).to.equal(403);
  });

  it('should succeed on match between token and required scope', () => {
    let error;
    try {
      _assertScope({
        tokenScope: 'read', requiredScope: 'read'
      });
    } catch(e) {
      error = e;
    }
    expect(error).to.not.exist;
  });

  it('should succeed when token scope contains required scope', () => {
    let error;
    try {
      _assertScope({
        tokenScope: 'read write', requiredScope: 'read'
      });
    } catch(e) {
      error = e;
    }
    expect(error).to.not.exist;
  });

  it('should fail when token scope is a subset of required', () => {
    let error;
    try {
      _assertScope({
        tokenScope: 'read',
        requiredScope: 'write read'
      });
    } catch(e) {
      error = e;
    }
    expect(error).to.exist;
    expect(error.error).to.equal('insufficient_scope');
    expect(error.error_description).to
      .equal('Access denied for token scope "read".');
    expect(error.statusCode).to.equal(403);
  });

  it('should succeed when token scope matches (order-independent)', () => {
    let error;
    try {
      _assertScope({
        tokenScope: 'read write', requiredScope: 'write read'
      });
    } catch(e) {
      error = e;
    }
    try {
      _assertScope({
        tokenScope: 'read write', requiredScope: 'write read'
      });
    } catch(e) {
      error = e;
    }
    expect(error).to.not.exist;
  });
});


tests/.eslintrc.js
module.exports = {
  globals: {
    should: true,
    assertNoError: true
  },
  env: {
    node: true,
    mocha: true
  },
  extends: [
    'digitalbazaar',
    'digitalbazaar/jsdoc'
  ],
  ignorePatterns: ['node_modules/']
};


lib/authorizeAccessToken.js
/*!
 * Copyright (c) 2021-2023 Digital Bazaar, Inc. All rights reserved.
 */
import * as authorization from 'auth-header';
import {InsufficientScope, InvalidRequest, InvalidScope, InvalidToken}
  from '@interop/oauth2-errors';
import assert from 'assert-plus';
import {LruCache} from '@digitalbazaar/lru-memoize';
import noopLogger from './noopLogger.js';

/**
 * Authorizes an incoming request for a route (parses Authorization: header,
 * decodes access token, verifies it, checks the decoded token claims against
 * the provided scope, sets the `req.clientMetadata` property).
 *
 * @param {object} options - Options hashmap.
 * @param {string} options.requiredScope - Scope required for the route guarded
 *   by this middleware.
 * @see https://tools.ietf.org/html/rfc6749#section-3.3
 * @param {string[]} [options.validIssuers] - List of allowed issuers of
 *   access tokens (e.g. ['https://example.com']). If missing, issuer validation
 *   should be performed in the validateClaims() callback.
 * @param {Function} [options.verify] - Decoding and signature verification
 *   function (validates the algorithm and verifies the signature).
 *   Expected function signature: async ({token}) => claims.
 * @param {Function} [options.validateClaims] - Optional custom validation
 *   callback with the signature `async ({claims})`, should throw appropriate
 *   errors.
 * @param {object} [options.logger] - Logger.
 * @param {object} [options.claimsCache] - Optional `lru-memoize` instance for
 *   storing decoded token claims). By default, the cache will be initialized
 *   with a max of 100 items, and an expiration of 5 seconds.
 * @param {Function} [options.decorateError] - Optional callback to customize
 *   the error response body (by default, uses the 3 fields from
 *   https://datatracker.ietf.org/doc/html/rfc6749#section-5.2).
 *
 * @returns {Function} Returns an Express.js middleware route handler.
 */
export function authorizeAccessToken({
  requiredScope, validIssuers, verify, validateClaims, decorateError,
  logger = noopLogger, claimsCache = new LruCache({max: 100, maxAge: 5000})
}) {
  assert.string(requiredScope, 'options.requiredScope');
  assert.optionalArrayOfString(validIssuers, 'options.validIssuers');
  assert.func(verify, 'options.verify');
  assert.optionalFunc(validateClaims, 'options.validateClaims');
  assert.optionalFunc(decorateError, 'options.decorateError');

  // Returns an Express middleware handle
  return async (req, res, next) => {
    try {
      // Extract access token from Authorization: header
      const {clientAccessToken} = _tokenFromHeader({
        authorizationHeader: req.get('authorization')
      });

      // Decode claims, verify signature
      let claims;
      try {
        (claims = await claimsCache.memoize({
          key: clientAccessToken,
          fn: async () => verify({token: clientAccessToken})
        }));
      } catch(e) {
        logger.debug('Error verifying access token.',
          {token: clientAccessToken, error: e});
        throw new InvalidToken({
          description: 'Error verifying the access token.'
        });
      }

      const {iss, exp, scope} = claims;

      // simple comparison, no leeway / clock skew
      _assertExpiration({exp});

      if(validIssuers) {
        _assertIssuer({iss, validIssuers});
      }

      if(validateClaims) {
        await validateClaims({claims});
      }

      _assertScope({tokenScope: scope, requiredScope, logger});

      // Success, set the decoded and validated `clientMetadata` property
      req.clientMetadata = {...claims};
    } catch(error) {
      return _handleError({error, logger, decorateError, res});
    }
    next();
  };
}

/**
 * Checks that the access token scope matches the required API endpoint's
 * scope, and also matches the scope granted to the client on registration.
 *
 * Scope is a whitespace-separated unordered list of substrings. Example valid
 * scopes:
 * - 'my.custom.scope' -- a single scope.
 * - 'read write' and 'write read' are equivalent, with an implicit AND.
 *
 * @throws {InvalidScope} If scope is missing or not a string.
 * @throws {InsufficientScope} If scopes do not match.
 *
 * @see https://tools.ietf.org/html/rfc6749#section-3.3
 *
 * @param {object} options - Options hashmap.
 * @param {string} options.tokenScope - Scope extracted from access_token.
 * @param {string} options.requiredScope - Scope required for a particular
 *   api endpoint.
 * @param {object} options.logger - Optional logger object.
 */
export function _assertScope({
  tokenScope, requiredScope, logger = noopLogger
} = {}) {
  if(typeof tokenScope !== 'string') {
    throw new InvalidScope(
      {description: `Invalid scope "${tokenScope}".`});
  }

  // Handle simple case (when there's a single scope or exact match)
  if(tokenScope === requiredScope) {
    return;
  }

  const tokenScopes = tokenScope.split(' ');
  const requiredScopes = requiredScope.split(' ');

  // Every required scope must be included in the token scope
  if(!requiredScopes.every(r => tokenScopes.includes(r))) {
    logger.debug(`Access denied for token scope "${tokenScope}". ` +
      `Scope required: "${requiredScope}".`);
    throw new InsufficientScope(
      {description: `Access denied for token scope "${tokenScope}".`});
  }
}

/**
 * Extracts the client access token from an incoming 'Authorization:` header.
 *
 * @example
 * // Incoming request has header: `Authorization: Bearer abcd12345`
 * _tokenFromHeader({authorizationHeader: req.get('authorization)};
 * // abcd12345
 *
 * @param {string} authorizationHeader - `authorization:` header value.
 *
 * @returns {string} Returns the extracted client access token.
 */
function _tokenFromHeader({authorizationHeader}) {
  let auth;
  try {
    // parse throws if the authorization header is missing
    auth = authorization.parse(authorizationHeader);
  } catch(e) {
    throw new InvalidRequest({
      description: 'Missing "authorization" header.'
    });
  }
  if(!(auth.scheme === 'Bearer' && auth.token)) {
    throw new InvalidRequest({
      description: 'Invalid "authorization" header.'
    });
  }
  return {clientAccessToken: auth.token};
}

function _assertExpiration({exp}) {
  // Date() works in milliseconds, `exp` is in seconds
  const now = Date.now() / 1000;
  if(exp < now) {
    throw new InvalidToken({
      description: 'Access token has expired.',
      name: 'ConstraintError'
    });
  }
}

function _assertIssuer({iss, validIssuers}) {
  if(!validIssuers.includes(iss)) {
    throw new InvalidToken({
      description: `Token issuer "${iss}" is not authorized.`,
      name: 'NotAllowedError'
    });
  }
}

function _handleError({error, logger, decorateError, res}) {
  logger.error('Authorization error:', {error});
  const {
    error: errorId, error_description: description, error_uri: uri, name
  } = error;

  const statusCode = error.statusCode || 400;
  const oauth2ErrorResponse = {
    error: errorId || 'invalid_request',
    error_description: description || error.message,
    error_uri: uri,
    name
  };

  if(decorateError) {
    decorateError({errorResponse: oauth2ErrorResponse});
  }

  res.set({
    'Cache-Control': 'no-store',
    Pragma: 'no-cache'
  });
  return res.status(statusCode).json(oauth2ErrorResponse);
}
